CFU, practice and worksheet for nested data and nested loops (#51)
Ada-Developers-Academy/core-unit-1: nested-data-and-nested-loops/practice-nested-data-and-nested-loops.md

    }
}

def get_school_name():
    pass
```
##### !end-placeholder
##### !tests
```python

import main
import unittest

class TestSchool(unittest.TestCase):
  def test_school_name(self):
    name = get_school_name()
    expected = "University of Washington"
    self.assertEqual(name, expected)
```
##### !end-tests
### !end-challenge
<!-- prettier-ignore-end -->

<!-- prettier-ignore-start -->
### !challenge
* type: code-snippet
* language: python3.6
* id: b54addb3-48b0-462a-9f6a-f9b749fba39f
* title: Practice Nested Data VI

##### !question
Modify `get_schools` to print the name and city of each school from `schools`.
##### !end-question

##### !placeholder
```python
schools = {
    "school 1": {
        "name": "Western Washington University",
        "city": "Bellingham"
    },
    "school 2": {
        "name": "University of Washington",
        "city": "Seattle"
    },
    "school 3": {
        "name": "Washington State University",
        "city": "Pullman"
    }
}

def get_schools():
    pass
```
##### !end-placeholder
##### !tests
```python

import main
import unittest
from unittest.mock import patch 

@patch('builtins.print')
class TestSchools(unittest.TestCase):

>   def test_schools(self, mock_print):

Reviewer: Same feedback.  Consider multiple cases (use of `global`) and `StringIO` to capture results with full fidelity.

    get_schools()
    mock_print.assert_called_with("Pullman")
```
##### !end-tests
### !end-challenge
<!-- prettier-ignore-end -->

<!-- prettier-ignore-start -->
### !challenge
* type: code-snippet
* language: python3.6
* id: 279c4012-341e-4461-a716-f0c093c7faba
* title: Practice Nested Data VII

##### !question

> It's Karoake night! Ada's karoake machine has a list of performers and Akira is up next with their second song. Modify `get_song_name()` to return `Mr. Brightside` from the `performers` list.

Reviewer: ```suggestion
It's Karaoke night! Ada's karaoke machine has a list of performers and Akira is up next with their second song. Modify `get_song_name()` to return `Mr. Brightside` from the `performers` list.
```

##### !end-question

##### !placeholder
```Python

performers = {
  "Rajan": ["I Want It That Way", "Jolene"],
  "Akira": ["Return of the Mack", "Mr. Brightside"],
  "Linh": ["Say My Name", "Since You Been Gone"]
}

def get_song_name():
    pass
```
##### !end-placeholder
##### !tests
```python

import main
import unittest

class TestSong(unittest.TestCase):

>   def test_song_name(self):

Reviewer: Same feedback.  Consider multiple cases (use of `global`).

    song = get_song_name()
    expected = "Mr. Brightside"
    self.assertEqual(name, expected)
```
##### !end-tests
### !end-challenge
<!-- prettier-ignore-end -->

<!-- prettier-ignore-start -->
### !challenge
* type: code-snippet
* language: python3.6
* id: b54addb3-48b0-462a-9f6a-f9b749fba39f
* title: Practice Nested Data VIII

##### !question

> Adie's want to keep record of all the songs performed during Karoake Night. Modify `get_songs()` to print all the songs from the `performers` list.

Reviewer: Minor rephrasing.
```suggestion
Let's make a record of all the songs performed during Karaoke Night! Modify `get_songs()` to print all the songs from the `performers` list.
```

##### !end-question

##### !placeholder
```Python

performers = {
  "Rajan": ["I Want It That Way", "Jolene"],
  "Akira": ["Return of the Mack", "Mr. Brightside"],
  "Linh": ["Say My Name", "Since You Been Gone"]
}

def get_songs():
    pass
```
##### !end-placeholder
##### !tests
```python

import main
import unittest
from unittest.mock import patch 

@patch('builtins.print')
class TestSchools(unittest.TestCase):

>   def test_schools(self, mock_print):

Reviewer: Same feedback.  Consider multiple cases (use of `global`) and `StringIO` to capture results with full fidelity.

# Practice: Nested Data and Nested Loops

## Directions

Complete all questions below.

## Practice

<!-- Question 1 -->
<!-- prettier-ignore-start -->
### !challenge
* type: short-answer
* id: 8mavqk
* title: Nested Data and Nested Loops
##### !question

What is the syntax to get `'6501 Island St. SE'` from `restaurants`?

```python
restaurants = [
    {
        "name": "Dough Zone",
        "cuisine": "Chinese",
        "address": "1299 Dough Ave N"
    },
    {
        "name": "Bingos",
        "cuisine": "Carribean",
        "address": "6501 Island St SE"
    },
    {
        "name": "Toyoda",
        "cuisine": "Japanese",
        "address": "12543 Sushi Way NE"
    }
]
```

Use `[]` syntax with double-quotes where needed.

##### !end-question
##### !answer

restaurants[1]["address"]

##### !end-answer
##### !hint

Use `[]` syntax with double-quotes where needed.

Examples:

- `some_list[0]`
- `some_dictionary["some key"]`

##### !end-hint
### !end-challenge
<!-- prettier-ignore-end -->

<!-- Question 2 -->
<!-- prettier-ignore-start -->
### !challenge
* type: code-snippet
* language: python3.6
* id: WfX3l1
* title: Nested Data and Nested Loops
### !question

> Implement `get_restaurant_addresses()`. This method takes in one parameter, `restaurants`. The method should return a list of strings, which are restaurant addresses.

Reviewer: This sounds more natural to me to describe the desired contents of the result, followed by the type information clarification.
```suggestion
Implement `get_restaurant_addresses()`. This method takes in one parameter, `restaurants`. The method should return a list of restaurant addresses, which are strings.
```

If the other ordering is desired, then I'd insert 'the' (the current phrasing makes it sound like 'restaurant address' is a type).
```suggestion
Implement `get_restaurant_addresses()`. This method takes in one parameter, `restaurants`. The method should return a list of strings, which are the restaurant addresses.
```

```
### !end-tests
### !explanation

An example of a working implementation:

```python
def get_restaurant_addresses(restaurants):
    addresses = []
    for restaurant in restaurants:
        addresses.append(restaurant["address"])
    return addresses
```

### !end-explanation
### !end-challenge
<!-- prettier-ignore-end -->

<!-- Question 3 -->
<!-- prettier-ignore-start -->
### !challenge
* type: short-answer
* id: 99WiYg
* title: Nested Data and Nested Loops
##### !question

What is the syntax to get `'Lizzie'` from `student_groups`?

```python
students = [
    ['Michelle', 'Beyonce', 'Kelly'],
    ['Luke', 'Leia', 'Han'],
    ['Gordon', 'Lizzie', 'Miranda']
]
```

Use `[]` syntax with double-quotes where needed.

##### !end-question
##### !answer

students[2][1]

##### !end-answer
##### !hint

Use `[]` syntax with double-quotes where needed.

Examples:

- `some_list[0]`
- `some_dictionary["some key"]`

##### !end-hint
### !end-challenge
<!-- prettier-ignore-end -->

<!-- Question 4 -->
<!-- prettier-ignore-start -->
### !challenge
* type: code-snippet
* language: python3.6
* id: UOoCrY
* title: Nested Data and Nested Loops
##### !question

> Implement `get_student_names()`. This method takes in one parameter, `student_groups`. The method should return a list of strings, which are student names.

Reviewer: Similar options as above
```suggestion
Implement `get_student_names()`. This method takes in one parameter, `student_groups`. The method should return a list of student names, which are strings.
```

or
```suggestion
Implement `get_student_names()`. This method takes in one parameter, `student_groups`. The method should return a list of strings, which are the student names.
```

    assert no_names == []
```

##### !end-question
##### !tests
```python
import unittest
from main import *

class TestPython1(unittest.TestCase):
    def test_get_student_names(self):
        students = [
            ['Michelle', 'Beyonce', 'Kelly'],
            ['Luke', 'Leia', 'Han'],
            ['Gordon', 'Lizzie', 'Miranda']
        ]
        names = get_student_names(students)
        self.assertEqual(
            names, ['Michelle', 'Beyonce', 'Kelly', 'Luke', 'Leia', 'Han', 'Gordon', 'Lizzie', 'Miranda'])

    def test_get_student_names_no_names(self):
        no_names = get_student_names([])

        self.assertEqual(no_names, [])
```
##### !end-tests
### !explanation

An example of a working implementation:

```python
def get_student_names(student_groups):
    names = []
    for group in student_groups:
        for student in group:
            names.append(student)
    return names
```

### !end-explanation
### !end-challenge
<!-- prettier-ignore-end -->

<!-- Question 5 -->
<!-- prettier-ignore-start -->
### !challenge
* type: short-answer
* id: 613ccf11-e7e6-4fad-a1b1-6a962bcd8967
* title: Nested Data and Nested Loops
##### !question

> What is the syntax to get `'University of Washington'` from `schools`?

Reviewer: ```suggestion
What is the syntax to get `"University of Washington"` from `schools`?
```

```python
schools = {
    "school 1": {
        "name": "Western Washington University",
        "city": "Bellingham"
    },
    "school 2": {
        "name": "University of Washington",
        "city": "Seattle"
    },
    "school 3": {
        "name": "Washington State University",
        "city": "Pullman"
    }
}
```

Use `[]` syntax with double-quotes where needed.

##### !end-question
##### !answer

schools["school 2"]["name"]

##### !end-answer
##### !hint

Use `[]` syntax with double-quotes where needed.

Examples:

- `some_list[0]`
- `some_dictionary["some key"]`

##### !end-hint
### !end-challenge
<!-- prettier-ignore-end -->

<!-- Question 6 -->
<!-- prettier-ignore-start -->
### !challenge
* type: code-snippet
* language: python3.6
* id: b54addb3-48b0-462a-9f6a-f9b749fba39f
* title: Nested Data and Nested Loops
##### !question

> Implement `get_school_names()`. This method takes in one parameter, `schools`. The method should return a list of strings, which are school names.

Reviewer: Successful completion of this depends on an implementation detail of CPython 3.6 (formalized in Python 3.7--documented in a note at https://docs.python.org/3/library/stdtypes.html#typesmapping). I notice that the python editors in Learn are at Python 3.6.

Can we assume that's CPython (so that tests like this will pass)?

Can we increase the hosted version if there are language features we encounter that came _after_ 3.6? (Students have 3.9 installed on their machines if they went through installfest.)

Reviewer: Similar options as above
```suggestion
Implement `get_school_names()`. This method takes in one parameter, `schools`. The method should return a list of school names, which are strings.
```

or
```suggestion
Implement `get_school_names()`. This method takes in one parameter, `schools`. The method should return a list of strings, which are the school names.
```

            },
            "school 3": {
                "name": "Barkington State University",
                "city": "Pullbark"
            }
        }
        names = get_school_names(schools)
        self.assertEqual(
            names, ["Western Barkington University",
                    "University of Barkington", "Barkington State University"])

    def test_get_school_names_no_names(self):
        no_names = get_school_names({})

        self.assertEqual(no_names, [])
```
##### !end-tests
### !explanation

An example of a working implementation:

```python
def get_school_names(schools):
    names = []
    for school_num in schools:
        names.append(schools[school_num]["name"])
    return names
```

Another example of a working implementation:

```python
def get_school_names(schools):
    names = []
    for school_num, school in schools.items():
        names.append(school["name"])
    return names
```

### !end-explanation
### !end-challenge
<!-- prettier-ignore-end -->

<!-- Question 7 -->
<!-- prettier-ignore-start -->
### !challenge
* type: short-answer
* id: Q9vx6M
* title: Nested Data and Nested Loops
##### !question

It's karaoke night!

> What is the syntax to get `'Jolene'` from `karaoke_queue`?

Reviewer: ```suggestion
What is the syntax to get `"Jolene"` from `karaoke_queue`?
```

```python
karaoke_queue = {
  "Rajan": ["I Want It That Way", "Jolene"],
  "Akira": ["Return of the Mack", "Mr. Brightside"],
  "Linh": ["Say My Name", "Since You Been Gone"]
}
```

Use `[]` syntax with double-quotes where needed.

##### !end-question
##### !answer

karaoke_queue["Rajan"][1]

##### !end-answer
##### !hint

Use `[]` syntax with double-quotes where needed.

Examples:

- `some_list[0]`
- `some_dictionary["some key"]`

##### !end-hint
### !end-challenge
<!-- prettier-ignore-end -->

<!-- Question 8 -->
<!-- prettier-ignore-start -->
### !challenge
* type: code-snippet
* language: python3.6
* id: 1VNr86
* title: Nested Data and Nested Loops
##### !question

It's karaoke night!

> Implement `get_song_list()`. This method takes in one parameter, `karaoke_queue`. The method should return a list of strings, which are song titles.

Reviewer: Similar options as above
```suggestion
Implement `get_song_list()`. This method takes in one parameter, `karaoke_queue`. The method should return a list of song titles, which are strings.
```

or
```suggestion
Implement `get_song_list()`. This method takes in one parameter, `karaoke_queue`. The method should return a list of strings, which are the song titles.
```
